Separate find files utility

davidh-ssec opened the pull request:
Fix #82 

TODO:

 - [x] Finalize Scene interface for this version and next
 - [x] Add at least one unit test of new utility function

davidh-ssec commented on satpy/readers/__init__.py:
```diff
@@ -249,18 +249,108 @@ def read_reader_config(config_files):
     return reader_instance
 
 
+def configs_for_reader(reader=None, ppp_config_dir=None):
```
Why is that a separate package?

davidh-ssec commented on satpy/composites/abi.py:
```diff
@@ -37,6 +37,15 @@ def four_element_average(d):
     return np.ma.mean(d.reshape(new_shape), axis=(1, 3))
 
 
+def simulated_green(c01, c02, c03):
```
No, but it was a small change so I thought I'd slip it in in this PR.

davidh-ssec commented:
Current filtering cases that need to be handled consistently:

1. Filter filenames based on filename metadata including start and end time
2. Filter file handlers based on internal file (or data) start and end time (file handler properties)
3. Filter file handlers based on geographic area
4. Filter file handlers based on file metadata (`fh.metadata` dictionary)

So 2, 3, and 4 all require opening the files first. I believe @mraspaud and @pnuu and I discussed this on Slack, but to implement 2, 3, and 4 requires the user to do:

```
scn = Scene(..., reader_kwargs={'metadata': {'start_time': datetime(...)}})
```

Or similar. Is that too ugly? Because of how flexible this could be I don't think it makes sense to use `**kwargs` collection to the `Scene` and have that pass it along to the file handler filtering.

Another option for 2, 3, and 4 is to separate out the file handler filtering to a separate `Scene` method like `filter_data_sources` or something similar. The file handler filtering already has to create the complete file handler object so this isn't wasting processing time as far as that is considered.

Lastly, besides filtering by area, start time, and end time, is there really other metadata that someone would filter by that couldn't be taken from the filename? I suppose there is always that chance. Plus this does allow for some really fancy filtering later on. For example if the file handler wanted to do a "quality control" check on `__init__` the user could then filter out all bad files `scn.filter_data_sources(quality_threshold=0.8)`.

Thoughts @pnuu @mraspaud @adybbroe?

davidh-ssec commented:
Additionally, I think the `Scene.__init__` parameters would eventually (after all deprecated features are removed) look like this:

```
def __init__(filenames, reader=None, ppp_config_dir=None, reader_kwargs=None):
```

mraspaud commented:
I would rather have the file filtering from file data allowed within the __init__. Is `reader_kwargs` going to be used for more than filtering, or could we rename this to eg `filtering_parameters` ? That would make it clearer.

Regarding netcdf saving, what I do now is just `scn.save_datasets(write='cf', filename='bla.nc')`. You want to make a shortcut for that ?

davidh-ssec commented:
The `reader_kwargs` could include other stuff and do in a few cases. My only example that I have right now is the NUCAPS reader where I have a `mask_surface` and a `mask_quality` keyword argument to do additional masking and filtering of the output data. I *think* these could be theoretically implemented as modifiers, but it's not how I've been thinking of it.

For NetCDF file, not necessarily a shortcut. The only difference (depending on how you implemented it) is that `scn.info` would be used as global attributes. I guess as long as `save_datasets` allows for global attributes in a keyword argument that there is no benefit to `scn.save(...)`.

mraspaud commented:
Ok, so should we make the `filtering_parameters` separate for clarity ?

davidh-ssec commented:
I'm ok with that since I'm not a huge fan of that double dict thing. Only odd gotcha with doing that would be if any of the filtering parameters should also be metadata of the reader...or I guess it just internally becomes metadata of the reader?

`Scene(..., filtering_parameters={...}) == Scene(..., reader_kwargs={metadata: {...}})`?

davidh-ssec commented:
@mraspaud I started doing the `filter_parameters` changes (chose that instead of `filtering_parameters` and started to find some unnecessary code. First and maybe most importantly is that I removed the `sensor` keyword from the regular reader loading during `Scene.__init__`. This means that the `Scene` requires a list of filenames and the reader names used to read them. There is no check for `sensor` and there is no automatic reader finding. I also removed the `base_dir` keyword since it shouldn't be used anymore. This simplified the `ReaderFinder` class enough that I just turned it in to a function.

I started to consider moving all of the filename filtering stuff out of the reader base classes but then realized that it would require duplicating the config reading and stops subclasses from customizing any filename filtering.

Lastly, I'm wondering if we should remove the `start_time`, `end_time`, and `area` keywords as individual keywords to the readers and instead make them part of `filter_parameters` or other. They don't seem to be used (even by the custom `mipp_xrit` reader) for anything except filename or file handler filtering.

Although this is "aggressive" as far as removing features from the `Scene` init, I think it simplifies the init a ton so that new users are scared away by the number of options.

pnuu commented:
Sounds good, although I haven't looked at the implementation. The only question arises from the `sensor`; are there any data formats that have several instruments in the same file? That would mean that the filename pattern can't be used to guess the reader. But then again, in those cases the reader should(?) be given explicitly by the user, right?.

mraspaud commented:
Works for me. I like putting all relevant keywords to `filtering_parameters`, even `sensor`. Having filtering mixins is fine by me if it make the code more concise/simpler.

Just a reminder though that we need to provide an externally available function to fetch the filenames, so that the `base_dir` behaviour could be reproduced if needed by the user.

davidh-ssec commented:
@pnuu Several sensors data in one file: Not that we support right now but I wouldn't be surprised if some exist. When you think of all the level 2 and higher level processing where collocation has to happen I could see people putting multiple in one.

@mraspaud Regarding `base_dir`, which functionality are you talking about?

mraspaud commented:
I was just refering to having a drop in replacement for `Scene(base_dir=...)` like eg:
`Scene(find_files(base_dir=...), ...)`

mraspaud commented on satpy/readers/yaml_reader.py:
```diff
@@ -167,6 +160,7 @@ def select_files_from_directory(self, directory=None):
         for pattern in self.file_patterns:
             matching = glob.iglob(os.path.join(directory, globify(pattern)))
             filenames.extend(matching)
+            print(pattern, filenames)
```
Should there be a print statement here ?

davidh-ssec replied:
No... :(

mraspaud commented on satpy/scene.py:
```diff
@@ -44,63 +44,100 @@
+                                      Shortcut for `reader_kwargs['filter_parameters']`.
+            reader_kwargs (dict): Keyword arguments to pass to specific reader instances.
+            ppp_config_dir (str): The directory containing the configuration files for satpy.
+            base_dir (str): (DEPRECATED) The directory to search for files containing the
```
Should we have an alternative indicated in the DEPRECATED parentheses (like the others) ?

davidh-ssec replied:
Yes.

mraspaud commented on satpy/tests/test_readers.py:
```diff
@@ -117,68 +117,61 @@ def test_no_args(self):
+    #         ri = load_readers(reader='viirs_sdr')
+    #         self.assertListEqual(list(ri.keys()), ['viirs_sdr'])
+    #     finally:
+    #         os.remove(fn)
```
Why comments ?

davidh-ssec replied:
To remember that the tests used to exist and they should be reimplemented with the new interface.

mraspaud commented on satpy/tests/test_scene.py:
```diff
@@ -108,91 +96,68 @@ def test_init_with_ppp_config_dir(self):
-                    reader_kwargs=None,
-                    metadata={}
-                )
+    # TODO: Rewrite this test for the 'find_files_and_readers' function
```
Should this be done in this PR ?

davidh-ssec replied:
Yes...probably.

davidh-ssec commented:
After I fix the conflicts, good to merge @mraspaud ? Maybe one more bug fix release with current develop then merge? I see that `mipp_xrit.py` is conflicting now which worries me because I thought we said that file isn't needed any more. Right?

mraspaud commented:
Great job on the documentation, @davidh-ssec ! I can merge if you say the word.
